### ISSUE TITLE
remove S3 prefix from recursive downloads

### ISSUE THREAD

**Issue by mgk** [mgk/s3-plugin#6]
Recursive download behavior is different than the aws cli in that it appends the entire S3 object path to the `destDir`.

https://github.com/mgk/s3-plugin#s3download

`S3Download` could have an option to behave like aws cli and not include S3 object path as subtree of `destDir`

**Comment by NickMcAvoy** [comment]
I agree with the aim to replicate `cli` functionality. However, it creates a problem: the need to relocate the resources from the downloaded temporary directory onto a directory which will already exist if the task has already run. This shouldn't be a big deal, but to my knowledge Java provides no single method for recursively deleting or overwriting a directory. So when I perform a single recursive download, I see what I expect. Upon a second:

```
:downloadSerman FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':downloadSerman'.
> java.nio.file.FileAlreadyExistsException: /home/nick/projects/idmp-dashboard/target/serman
```

As far as I know, the only options are:
-  Use a third-party library, such as Apache Commons's FileUtils or Google Guava
-  Iterate manually over the destination directory if it exists and first delete everything.

Both of these seem undesirable.

**Comment by mgk** [comment]
Could you supply the details of what you did to produce the error above so I can reproduce the error?

- an example gradle file
- the gradle command you ran to produce error
- output of running with `gradle -s` which includes stack traces

I can avoid the directory move in the case where `pruneTree` is false, but I'd also like to understand and fix what you are running into.

**Comment by mgk** [comment]
Oh I see. If you do a recursive download with the same `destDir` the second one fails as above. This is not a problem that the s3-plugin should try to solve. The s3-plugin should complain nicely and exit if `destDir` exists. Then you can have your download task depend on another task that removes or moves the destDir.

**Comment by NickMcAvoy** [comment]
I understand that perspective but I don't quite agree, though I do agree that just wiping the destination directory as I indicated is probably a bad idea.
-  If the guiding vision for the plugin is replicating the functionality of the `cli` client, the client has no problem downloading into an existing directory and overwriting any of the specific files in the target that are present in the source tree.
-  Idempotency would be a useful property for the task to have, that it makes no difference whether the task is run 1 or N times. Off the top of my head, as a rule Gradle tasks seem to be idempotent; if I run `jar` four times I still just get a jar in my build directory. Otherwise, keeping track of the state introduces a lot of overhead.
-  Symmetry with the plugin's other tasks would also dictate overwriting. For example, I have not tested, but I assume a recursive upload to `s3` will proceed whether or not the target exists. Also, download task of a single file does not fail if a target file exists on disk already.

So instead of overwriting, I would recommend that the plugin merge the downloaded temporary directory with the target directory if it already exists. To be clear, I mean if I am downloading the following tree from `s3:`

```
abc/def.txt
```
But at the `destDir` the following exists:
```
abc/def.txt
abc/ghi.txt
```

Then I think the plugin should overwrite `def.txt` and leave `ghi.txt` present, and the task should succeed.

Nick McAvoy     [cid:image001.jpg@01D15756.FA2640F0]
Senior Consultant
Insurance Technology Services, Inc.
mobile: +82 (010) 8459-5265
skype: nicholas.mcavoy
nick.mcavoy@insurancetechnologyservices.com<https://west.exch027.serverdata.net/owa/redir.aspx?SURL=REQOFnL8avSGKCgz4WfSUavM4iekiU2dQHbwNjnt_QX9zaHLB1rTCG0AYQBpAGwAdABvADoAbgBpAGMAawAuAG0AYwBhAHYAbwB5AEAAaQBuAHMAdQByAGEAbgBjAGUAdABlAGMAaABuAG8AbABvAGcAeQBzAGUAcgB2AGkAYwBlAHMALgBjAG8AbQA.&URL=mailto%3anick.mcavoy%40insurancetechnologyservices.com>
www.insurancetechnologyservices.com<https://west.exch027.serverdata.net/owa/redir.aspx?SURL=QmMmeYAdhCRW2D48ouWUYpSspLAul-PSMYYoHOv9WGv9zaHLB1rTCGgAdAB0AHAAOgAvAC8AdwB3AHcALgBpAG4AcwB1AHIAYQBuAGMAZQB0AGUAYwBoAG4AbwBsAG8AZwB5AHMAZQByAHYAaQBjAGUAcwAuAGMAbwBtAA..&URL=http%3a%2f%2fwww.insurancetechnologyservices.com>
________________________________

**Comment by mgk** [comment]
Thanks for your thoughts on this. While looking into different ways to do directory moves and copies  I realized the s3-plugin shouldn't be doing this at all. The Gradle Copy and Sync tasks already exist for this purpose. Those built in tasks are very flexible, support up to date checking, and are thoroughly tested.

The goal for the plugin is to have basic S3 upload and download behavior. I wanted it do do one thing and do it well.

So instead I added an example of using the Gradle Copy task to achieve the same affect as the basic pruning that came to my mind when I logged this issue. Staging files under the build tree somewhere and then using a Copy or Archive task to deploy or publish those files is more of a Gradle best practice anyway.

**Comment by mgk** [comment]
Closing this one, but feel free to comment here or open a new issue if there is a use case that is not handled with the combination of the S3Download and built in gradle tasks.

**Comment by NickMcAvoy** [comment]
Thanks for your attention to this. I think you're right that there's no good reason to write code to move the directory when there are fundamental Gradle tasks for doing that. 

Leveraging a `Copy` task is simple enough, but all the same, what about providing a baked-in way to do this?

**Comment by mgk** [comment]
> Leveraging a Copy task is simple enough, but all the same, what about providing a baked-in way to do this?

I was thinking along those lines but didn't follow through because I didn't see a default behavior that I thought most people would like/expect. And I did not want to go down the slippery slope of providing various options which would be passed through to `Copy`.

But I'm open to this idea if it can be kept simple enough. Could you say more, maybe with a  example about what behavior you'd like here?
